### PR TITLE
feat: yarn workspace command

### LINKED ISSUE #594: Yarn Workspace Command for Metapool Migration
I should of included this as part of the original PR.

### PR DESCRIPTION
**Pull request by hashedMae**
Resolves #594 
Simplify execution of metapool migration. 
Populate .env with `RPC_URL` and `PRIVATE_KEY`
In terminal run `yarn workspace @ubiquity/contracts migrate:metapool`

Breakdown of all components of the Solidity Script
Contracts it's utilizing: 
[uAD:3CRV V2 Metapool 
](https://etherscan.io/address/0x20955cb69ae1515962177d164dfc9522feef567e)

[uAD:3CRV V3 Metapool](https://etherscan.io/address/0x9558b18f021FC3cBa1c9B777603829A42244818b)

[3CRV Pool Token](https://etherscan.io/address/0x6c3f90f043a72fa612cbac8115ee7e52bde6e490)

[Ubiquity Algorithmic Dollar](https://etherscan.io/address/0x0F644658510c95CB46955e55D7BA9DDa9E9fBEc6)

[BondingV2](https://etherscan.io/address/0xC251eCD9f1bD5230823F9A0F99a44A87Ddd4CA38)

[Ubiquity Algorithmic Dollar Manager](https://etherscan.io/address/0x4DA97a8b831C345dBe6d16FF7432DF2b7b776d98)

`dollarPreBalance` uAD Tokens in the admin wallet before migration takes place

`metaBalance` V2 Metapool LP tokens currently in BondingV2

Calls `sendDust` to move V2 Metapool LP Tokens from BondingV2 to admin address

Calls `remove_liquidity` on V2 Metapool using `metaBalance`

`deposit` the lesser of the balances of `3CRV` or `uAD` tokens in admin address after withdrawal

Approve and deposits tokens to V3 Metapool using `deposit` as the amount for each to ensure the pool is in balance. The V3 Metapool LP Tokens are sent directly to BondingV2.

Require statement to ensure the V3 Metapool LP balance of BondingV2 is the same as how many LP tokens were just minted. 

Deploys a new `TWAPOracle` using the V3 Metapool.

Calls `setTwapOracleAddress()` in Ubiquity Dollar Manager.

`secondDeposit` The amount of uAD tokens remaining in admin address minus `dollarPreBalance`.

Calls `add_liquidty` on V3 Metapool with `secondDeposit` as the amount for uAD and 0 as the amount for 3CRV. V3 Metapool LP Tokens are again sent directly to BondingV2.

Another require statement to ensure the newly minted LP tokens are in BondingV2.

Calls `setStableSwapMetaPoolAddress()` in Ubiquity Dollar Manager

Require statement to ensure new `TWAPOracle` address was set correctly in Ubiquity Dollar Manager.

Require statement to ensure new `Metapool` address was set correctly in Ubiquity Dollar Manager.

### CODE REVIEW

**Comment by hashedMae** [comment]
I haven't been able to sim running this script directly, as I don't have the admin private key and am more comfortable not knowing it. The calls and order was copied to test files and was able to execute there successfully. I don't anticipate any hiccups, but if it doesn't run successfully I may have to simulate by deploying a mirror environment on anvil with the test accounts.

**Comment by 0x4007** [comment]
> I haven't been able to sim running this script directly, as I don't have the admin private key and am more comfortable not knowing it. The calls and order was copied to test files and was able to execute there successfully. I don't anticipate any hiccups, but if it doesn't run successfully I may have to simulate by deploying a mirror environment on anvil with the test accounts.

Shouldn't you be able to easily impersonate the admin wallet to test?

**Comment by hashedMae** [comment]
> > I haven't been able to sim running this script directly, as I don't have the admin private key and am more comfortable not knowing it. The calls and order was copied to test files and was able to execute there successfully. I don't anticipate any hiccups, but if it doesn't run successfully I may have to simulate by deploying a mirror environment on anvil with the test accounts.
> 
> Shouldn't you be able to easily impersonate the admin wallet to test?

Using `forge test` yes, but not `forge script`. That's why I wrote the test files.

**Comment by 0x4007** [comment]
Hey @zgorizzo69 just following up on a review for this. Thanks!

File: packages/contracts/scripts/task/dollar/MigrateMetapool.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+if (!fs.existsSync(envPath)) {
+  throw new Error("Env file not found");
+}
+const env = loadEnv(envPath);
+
+const rpcUrl = env.rpcUrl as string;
+
+const migrateMetapool = async () => {
+  console.log("Running Solidity Script to Migrate Metapool Funds");
+  spawnSync("forge", ["script", "scripts/dollar/MigrateMetapool.s.sol", "--fork-url", rpcUrl, "--broadcast"]);
```
Comment (zgorizzo69) [inline packages/contracts/scripts/task/dollar/MigrateMetapool.ts]: where is that file ?

**Comment by zgorizzo69** [comment]
I ran anvil and this command `yarn migrate:metapool`
as there is no throw on error I was not able to see if it worked so I ran this command instead 
`forge script scripts/task/dollar/MigrateMetapool.s.sol  --fork-url http://localhost:8545 --broadcast `  which is basically what 
`yarn migrate:metapool` does and there are some errors like 
- `dollarPreBalance` using admin before it is defined
- `v3Metapool.add_liquidity([secondDeposit, uint256[0]],..` instead of `v3Metapool.add_liquidity([secondDeposit, uint256(0)],`
-  `manager.stableSwapMetaPoolAddress == address(v3Metapool),` instead of  `manager.stableSwapMetaPoolAddress() == address(v3Metapool),`
- etc..

I suggest you do the following 
- rename `migrateFunds` to `MigrateFunds` cos all className start with an uppercase
- refactor `MigrateFunds` to extract the logic inside internal functions and just call them inside run()
- now you can update the script test  `MigrateMetapool.t.sol` (also probably best to rename this class to MigrateMetapoolTest)  and test the MigrateFunds internal functions logic

**Review by zgorizzo69 — CHANGES_REQUESTED** [review]

**Review by zgorizzo69 — APPROVED** [review]

**Comment by 0x4007** [comment]
@hashedMae @zgorizzo69 can one (or both) of you walk me through executing this assuming that its ready?

**Comment by hashedMae** [comment]
> @hashedMae @zgorizzo69 can one (or both) of you walk me through executing this assuming that its ready?

Populate .env with RPC_URL and PRIVATE_KEY
In terminal run yarn workspace @ubiquity/contracts migrate:metapool

**Pull request merged**